map/chart_generator.cpp

```diff
@@ -0,0 +1,236 @@
+  uint32_t constexpr kBPP = 4;
+
+  using TBlender = BlendAdaptor<agg::rgba8, agg::order_rgba>;
+  using TPixelFormat = agg::pixfmt_custom_blend_rgba<TBlender, agg::rendering_buffer>;
+  using TBaseRenderer = agg::renderer_base<TPixelFormat>;
+  using TPrimitivesRenderer = agg::renderer_primitives<TBaseRenderer>;
+  using TSolidRenderer = agg::renderer_scanline_aa_solid<TBaseRenderer>;
+  using TPath = agg::poly_container_adaptor<vector<m2::PointD>>;
+  using TStroke = agg::conv_stroke<TPath>;
+
+  agg::rendering_buffer renderBuffer;
+  TPixelFormat pixelFormat(renderBuffer, agg::comp_op_src_over);
+  TBaseRenderer baseRenderer(pixelFormat);
+
+  frameBuffer.assign(width * kBPP * height, 0);
+  renderBuffer.attach(&frameBuffer[0], static_cast<unsigned>(width),
+                      static_cast<unsigned>(height), static_cast<int>(width * kBPP));
+  baseRenderer.reset_clipping(true);
+  unsigned op = pixelFormat.comp_op();
+  pixelFormat.comp_op(agg::comp_op_src);
+  baseRenderer.clear(kBackgroundColor);
+  pixelFormat.comp_op(op);
+
+  agg::rasterizer_scanline_aa<> rasterizer;
+  rasterizer.clip_box(0, 0, width, height);
+
+  // Polygon under chart line.
+  agg::path_storage underChartGeometryPath;
+  underChartGeometryPath.move_to(geometry.front().x, static_cast<double>(height));
+  for (auto const & p : geometry)
+    underChartGeometryPath.line_to(p.x, p.y);
+  underChartGeometryPath.line_to(geometry.back().x, static_cast<double>(height));
+  underChartGeometryPath.close_polygon();
+
+  agg::conv_curve<agg::path_storage> curve(underChartGeometryPath);
+  rasterizer.add_path(curve);
+  agg::scanline32_p8 scanline;
+  agg::render_scanlines_aa_solid(rasterizer, scanline, baseRenderer, kCurveColor);
+
+  // Chart line.
+  TPath path_adaptor(geometry, false);
+  TStroke stroke(path_adaptor);
+  stroke.width(kLineWidthPxl);
+  stroke.line_cap(agg::round_cap);
+  stroke.line_join(agg::round_join);
+
+  rasterizer.add_path(stroke);
+  agg::render_scanlines_aa_solid(rasterizer, scanline, baseRenderer, kLineColor);
+}
+
+bool GenerateChart(uint32_t width, uint32_t height, vector<double> const & distanceDataM,
+                   feature::TAltitudes const & altitudeDataM, bool day,
+                   vector<uint8_t> & frameBuffer)
+{
+  if (distanceDataM.size() != altitudeDataM.size())
+  {
+    LOG(LERROR, ("The route is in inconsistent state. Size of altitudes is", altitudeDataM.size(),
+                 ". Number of segment is", distanceDataM.size()));
+    return false;
+  }
+
+  frameBuffer.clear();
+  if (altitudeDataM.empty())
+    return true;
+
+  vector<double> uniformAltitudeDataM;
+  NormalizeChartData(distanceDataM, altitudeDataM, width, uniformAltitudeDataM);
+  if (uniformAltitudeDataM.empty())
+    return true;
+
+  vector<double> yAxisDataPxl;
+  GenerateYAxisChartData(height, 1.0 /* minMetersPerPxl */, uniformAltitudeDataM, yAxisDataPxl);
+
+  size_t const uniformAltitudeDataSz = yAxisDataPxl.size();
+  if (uniformAltitudeDataSz <= 1)
```

Comment:
Now I return false if uniformAltitudeDataSz is less than one. And I do not touch frame buffer in that case.